Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,905 @@
+    simp only [Subtype.forall, Set.mem_setOf_eq] at hσ hτ
+    intro c hc
+    rw [hσ, hτ]
+  inv_mem' := by
+    intro σ hσ
+    simp only [Subtype.forall, Set.mem_setOf_eq] at hσ ⊢
+    intro c hc
+    rw [← hσ]
+    · simp only [Finset.coe_mem, Subtype.coe_eta, apply_inv_self]
+    · simp only [Finset.coe_mem]
+
+variable {g} in
+theorem mem_range_toPermHom'_iff {τ : Perm g.cycleFactorsFinset} :
+    τ ∈ range_toPermHom' g ↔
+      ∀ c, (τ c : Perm α).support.card = (c : Perm α).support.card :=
+  Iff.rfl
+
+/-- `k : Subgroup.centralizer {g}` belongs to the kernel of `toPermHom g`
+iff it commutes with each cycle of `g` -/
+theorem mem_ker_toPermHom_iff (k : Subgroup.centralizer {g}) :
+    k ∈ (toPermHom g).ker ↔
+      ∀ c ∈ g.cycleFactorsFinset, Commute (k : Perm α) c := by
+  simp only [toPermHom, MonoidHom.mem_ker, DFunLike.ext_iff]
+  simp only [coe_one, id_eq, Subtype.forall]
+  apply forall₂_congr
+  intro c hc
+  simp only [MulAction.toPermHom_apply, toPerm_apply, ← Subtype.coe_inj,
+    commute_iff_eq, centralizer_smul_def, InvMemClass.coe_inv,
+    mul_inv_eq_iff_eq_mul]
+
+end OnCycleFactors
+
+open OnCycleFactors Subgroup
+
+/-- A `Basis` of a permutation is a choice of an element in each of its cycles -/
+class Basis (g : Equiv.Perm α) where
+  /-- A choice of elements in each cycle -/
+  (toFun : g.cycleFactorsFinset → α)
+  /-- For each cycle, the chosen element belongs to the cycle -/
+  (mem_support_self' : ∀ (c : g.cycleFactorsFinset), toFun c ∈ (c : Perm α).support)
+
+instance (g : Perm α) :
+    DFunLike (Basis g) (g.cycleFactorsFinset) (fun _ => α) where
+  coe a := a.toFun
+  coe_injective' a a' _ := by cases a; cases a'; congr
+
+namespace Basis
+
+theorem nonempty (g : Perm α) : Nonempty (Basis g) := by
+  have (c : g.cycleFactorsFinset) : (c : Perm α).support.Nonempty :=
+    IsCycle.nonempty_support (mem_cycleFactorsFinset_iff.mp c.prop).1
+  exact ⟨fun c ↦ (this c).choose, fun c ↦ (this c).choose_spec⟩
+
+variable (a : Basis g)
+
+theorem mem_support_self (c : g.cycleFactorsFinset) :
+    a c ∈ (c : Perm α).support := a.mem_support_self' c
+
+theorem injective : Function.Injective a := by
+  intro c d h
+  rw [← Subtype.coe_inj]
+  apply g.cycleFactorsFinset_pairwise_disjoint.eq c.prop d.prop
+  simp only [Disjoint, not_forall, not_or]
+  use a c
+  conv_rhs => rw [h]
+  simp only [← Perm.mem_support, a.mem_support_self c, a.mem_support_self d, and_self]
+
+theorem cycleOf_eq (c : g.cycleFactorsFinset) :
+    g.cycleOf (a c) = c :=
+  (cycle_is_cycleOf (a.mem_support_self c) c.prop).symm
+
+variable (τ : range_toPermHom' g)
+
+/-- The function that will provide a right inverse `toCentralizer` to `toPermHom` -/
+noncomputable def newK (x : α) : α := by
+  if hx : g.cycleOf x ∈ g.cycleFactorsFinset
+  then
+    let h := mem_support_cycleOf_iff.mp (a.mem_support_self ⟨g.cycleOf x, hx⟩)
+    exact (g ^ h.1.symm.choose) (a ((τ : Perm g.cycleFactorsFinset) ⟨g.cycleOf x, hx⟩))
```

Comment:
It seems like `h` is just expressing that `x` and `a x` are in the same cycle of `g`. I haven't thought this through, but I wonder if it's possible to streamline things by turning `a` into a function `α → α`, which would allow for a more uniform definition of `newK`?

Reply:
(Maybe this is more tricky than I thought, since `τ` is still a permutation of the cycle factors).

Reply:
plus that `x` is in the support of `g`. What would have been easier is to have access to the orbits of `g`, while `g.cycleFactorsFinset` only gives access to the nontrivial ones.

Reply:
A function `a : α → α` such that for every `x`, `SameCycle x (a x)` and `a x = a y ↔ SameCycle x y` would probably be equally usable for `newK`, but for the final formula, we need a product over `g.cycleFactorsFinset`, etc.